llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp

```diff
@@ -2653,8 +2653,15 @@ static Instruction *foldFNegIntoConstant(Instruction &I, const DataLayout &DL) {
   // Fold negation into constant operand.
   // -(X * C) --> X * (-C)
   if (match(FNegOp, m_FMul(m_Value(X), m_Constant(C))))
-    if (Constant *NegC = ConstantFoldUnaryOpOperand(Instruction::FNeg, C, DL))
+    if (Constant *NegC = ConstantFoldUnaryOpOperand(Instruction::FNeg, C, DL)) {
+      if (match(C, m_AnyZeroFP()) && I.getFastMathFlags().noInfs()) {
+        return BinaryOperator::CreateFMulFMF(
+            X, NegC,
+            I.getFastMathFlags() &
+                cast<FPMathOperator>(I.getOperand(0))->getFastMathFlags());
```

Comment:
Could we do it this way?
```cpp
      FastMathFlags FMF = cast<FPMathOperator>(I.getOperand(0))->getFastMathFlags();
      if (I.getFastMathFlags().noInfs() && !FMF.noInfs()) {
        return BinaryOperator::CreateFMulFMF(X, NegC, FMF);
      }
```

Reply:
Ok, we need to check all alive2 cases and then generate the flags.

Reply:
alive2: https://alive2.llvm.org/ce/z/DHmA7a
I apologize for the delay; I was quite busy. Based on my observation, I came up with this combination: if we have `nnan` in any instruction, we can propagate flags to our transformation.

```cpp
FastMathFlags NF1 = F1 & F2;
FastMathFlags NF2 = F1 | F2;

if (NF2.noNaNs()) {
    return F2;
}

if (NF1.any() || (F1.none() && F2.none())) {
    return NF1;
}

if (F1.all() || F2.all()) {
    return F2;
}

return NF1;
```

Let me know what you think.

Reply:
`NF1.any() || (F1.none() && F2.none())` case is equivalent to just NF1.any.

Not sure why you need the all || all fixup either

Reply:
I think ninf is the only problematic flag, since ninf on the fneg does not imply fneg on fmul, and fmul inf * 0 = nan. 

So you need something like

Flags = FnegFlags | FMulFlags
Flags.ninf &= FmulFlags.ninf